File: src/test/java/com/alibaba/ttl/user_test/TransmittableThreadLocal_Transmitter_UserTest.java
package com.alibaba.ttl.user_test;

import com.alibaba.ttl.TransmittableThreadLocal;
import org.junit.AfterClass;
import org.junit.Test;

import java.util.Date;
import java.util.concurrent.*;

import static com.alibaba.ttl.Utils.expandThreadPool;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * Test {@link TransmittableThreadLocal.Transmitter} from user code(different package)
 */
public class TransmittableThreadLocal_Transmitter_UserTest {
    private static final String PARENT = "parent: " + new Date();
    private static final String CHILD = "child: " + new Date();

    private static final ExecutorService executorService = Executors.newFixedThreadPool(3);

    static {
        expandThreadPool(executorService);
    }

    @AfterClass
    public static void afterClass() throws Exception {
        executorService.shutdown();
        executorService.awaitTermination(100, TimeUnit.MILLISECONDS);
        if (!executorService.isTerminated()) fail("Fail to shutdown thread pool");
    }

    @Test
    public void test_crr() throws Exception {
        final TransmittableThreadLocal<String> ttl = new TransmittableThreadLocal<String>();
        ttl.set(PARENT);

        final Object capture = TransmittableThreadLocal.Transmitter.capture();

        Future<?> future = executorService.submit(new Runnable() {
            @Override
            public void run() {
                ttl.set(CHILD);

                Object backup = TransmittableThreadLocal.Transmitter.replay(capture);

                assertEquals(PARENT, ttl.get());

                TransmittableThreadLocal.Transmitter.restore(backup);

                assertEquals(CHILD, ttl.get());
            }
        });

        assertEquals(PARENT, ttl.get());

        future.get(100, TimeUnit.MILLISECONDS);

        assertEquals(PARENT, ttl.get());
    }

    @Test
    public void test_runCallableWithCaptured() throws Exception {
        final TransmittableThreadLocal<String> ttl = new TransmittableThreadLocal<String>();
        ttl.set(PARENT);

        final Object capture = TransmittableThreadLocal.Transmitter.capture();

        Future<?> future = executorService.submit(new Runnable() {
            @Override
            public void run() {
                ttl.set("child");
                try {
                    TransmittableThreadLocal.Transmitter.runCallableWithCaptured(capture, new Callable<Object>() {
                        @Override
                        public Object call() throws Exception {
                            assertEquals(PARENT, ttl.get());
                            return ttl.get();
                        }
                    });
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        });

        assertEquals(PARENT, ttl.get());

        future.get(100, TimeUnit.MILLISECONDS);

        assertEquals(PARENT, ttl.get());
    }
}

File: src/test/java/com/alibaba/ttl/user_api_test/TransmittableThreadLocal_Transmitter_UserTest.kt
package com.alibaba.ttl.user_api_test

import com.alibaba.expandThreadPool
import com.alibaba.support.junit.conditional.BelowJava8
import com.alibaba.support.junit.conditional.ConditionalIgnoreRule
import com.alibaba.support.junit.conditional.ConditionalIgnoreRule.ConditionalIgnore
import com.alibaba.ttl.TransmittableThreadLocal
import org.junit.AfterClass
import org.junit.Assert.*
import org.junit.Rule
import org.junit.Test
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * Test [TransmittableThreadLocal.Transmitter] from user code(different package)
 */
class TransmittableThreadLocal_Transmitter_UserTest {
    @Rule
    @JvmField
    val rule = ConditionalIgnoreRule()

    @Test
    fun test_crr() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val capture = TransmittableThreadLocal.Transmitter.capture()

        val future = executorService.submit {
            ttl.set(CHILD)

            val backup = TransmittableThreadLocal.Transmitter.replay(capture)

            assertEquals(PARENT, ttl.get())

            TransmittableThreadLocal.Transmitter.restore(backup)

            assertEquals(CHILD, ttl.get())
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    @Test
    fun test_clear_restore() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val future = executorService.submit {
            ttl.set(CHILD)

            val backup = TransmittableThreadLocal.Transmitter.clear()

            assertNull(ttl.get())

            TransmittableThreadLocal.Transmitter.restore(backup)

            assertEquals(CHILD, ttl.get())
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    @Test
    @ConditionalIgnore(condition = BelowJava8::class)
    fun test_runSupplierWithCaptured() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val capture = TransmittableThreadLocal.Transmitter.capture()

        val future = executorService.submit {
            ttl.set("child")
            TransmittableThreadLocal.Transmitter.runSupplierWithCaptured(capture) {
                assertEquals(PARENT, ttl.get())
                ttl.get()
            }
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    @Test
    @ConditionalIgnore(condition = BelowJava8::class)
    fun test_runSupplierWithClear() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val future = executorService.submit {
            ttl.set("child")
            TransmittableThreadLocal.Transmitter.runSupplierWithClear {
                assertNull(ttl.get())
                ttl.get()
            }
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    @Test
    fun test_runCallableWithCaptured() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val capture = TransmittableThreadLocal.Transmitter.capture()

        val future = executorService.submit {
            ttl.set("child")
            try {
                TransmittableThreadLocal.Transmitter.runCallableWithCaptured(capture) {
                    assertEquals(PARENT, ttl.get())
                    ttl.get()
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    @Test
    fun test_runCallableWithClear() {
        val ttl = TransmittableThreadLocal<String>()
        ttl.set(PARENT)

        val future = executorService.submit {
            ttl.set("child")
            try {
                TransmittableThreadLocal.Transmitter.runCallableWithClear {
                    assertNull(ttl.get())
                    ttl.get()
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        assertEquals(PARENT, ttl.get())

        future.get(100, TimeUnit.MILLISECONDS)

        assertEquals(PARENT, ttl.get())
    }

    companion object {
        private val PARENT = "parent: " + Date()
        private val CHILD = "child: " + Date()

        private val executorService: ExecutorService = Executors.newFixedThreadPool(3).also { expandThreadPool(it) }

        @AfterClass
        @JvmStatic
        @Suppress("unused")
        fun afterClass() {
            executorService.shutdown()
            executorService.awaitTermination(100, TimeUnit.MILLISECONDS)
            if (!executorService.isTerminated) fail("Fail to shutdown thread pool")
        }
    }
}
